core/primitives/src/congestion_info.rs

```diff
@@ -57,18 +136,374 @@ impl Default for CongestionInfo {
+    // between the two. Even with precision errors, we cannot breach the
+    // boundaries.
+    return total.round() as u64;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_mix() {
+        assert_eq!(500, mix(0, 1000, 0.5));
+        assert_eq!(0, mix(0, 0, 0.3));
+        assert_eq!(1000, mix(1000, 1000, 0.1));
+        assert_eq!(60, mix(50, 80, 0.33));
+    }
+
+    #[test]
+    fn test_mix_edge_cases() {
+        // at `u64::MAX` we should see no precision errors
+        assert_eq!(u64::MAX, mix(u64::MAX, u64::MAX, 0.33));
+        assert_eq!(u64::MAX, mix(u64::MAX, u64::MAX, 0.63));
+        assert_eq!(u64::MAX, mix(u64::MAX, u64::MAX, 0.99));
+
+        // precision errors must be consistent
+        assert_eq!(u64::MAX, mix(u64::MAX - 1, u64::MAX, 0.25));
+        assert_eq!(u64::MAX, mix(u64::MAX - 255, u64::MAX, 0.25));
+        assert_eq!(u64::MAX, mix(u64::MAX - 1023, u64::MAX, 0.25));
+
+        assert_eq!(u64::MAX - 2047, mix(u64::MAX - 1024, u64::MAX, 0.25));
+        assert_eq!(u64::MAX - 2047, mix(u64::MAX - 1500, u64::MAX, 0.25));
+        assert_eq!(u64::MAX - 2047, mix(u64::MAX - 2047, u64::MAX, 0.25));
+        assert_eq!(u64::MAX - 2047, mix(u64::MAX - 2048, u64::MAX, 0.25));
+        assert_eq!(u64::MAX - 2047, mix(u64::MAX - 2049, u64::MAX, 0.25));
+        assert_eq!(u64::MAX - 2047, mix(u64::MAX - 3000, u64::MAX, 0.25));
+
+        assert_eq!(u64::MAX - 4095, mix(u64::MAX - 4000, u64::MAX, 0.25));
+    }
+
+    #[test]
+    fn test_clamped_f64_fraction() {
+        assert_eq!(0.0, clamped_f64_fraction(0, 10));
+        assert_eq!(0.5, clamped_f64_fraction(5, 10));
+        assert_eq!(1.0, clamped_f64_fraction(10, 10));
+
+        assert_eq!(0.0, clamped_f64_fraction(0, 1));
+        assert_eq!(0.0, clamped_f64_fraction(0, u64::MAX));
+
+        assert_eq!(0.5, clamped_f64_fraction(1, 2));
+        assert_eq!(0.5, clamped_f64_fraction(100, 200));
+        assert_eq!(0.5, clamped_f64_fraction(u64::MAX as u128 / 2, u64::MAX));
+
+        // test clamp
+        assert_eq!(1.0, clamped_f64_fraction(11, 10));
+        assert_eq!(1.0, clamped_f64_fraction(u128::MAX, 10));
+        assert_eq!(1.0, clamped_f64_fraction(u128::MAX, u64::MAX));
+    }
+
+    /// Default congestion info should be no congestion => maximally permissive.
+    #[test]
+    fn test_default_congestion() {
+        let inner_congestion_info = CongestionInfoV1::default();
+
+        assert_eq!(0.0, inner_congestion_info.memory_congestion());
+        assert_eq!(0.0, inner_congestion_info.incoming_congestion());
+        assert_eq!(0.0, inner_congestion_info.outgoing_congestion());
+        assert_eq!(0.0, inner_congestion_info.congestion_level());
+
+        let congestion_info = CongestionInfo::V1(inner_congestion_info);
+        assert_eq!(MAX_GAS_FORWARDING, congestion_info.outgoing_limit(0));
+        assert_eq!(MAX_TX_GAS, congestion_info.process_tx_limit());
```

Comment:
nit: It would be nice to have the same naming used for the constants and the limits e.g.
MAX_OUTGOING_GAS - outgoing_gas_limit()
MAX_TX_GAS - tx_gas_limit()